src/harps/io.jl

```diff
@@ -0,0 +1,171 @@
+    dict = read_metadata_from_fits(fn,fields=fields_to_save,fields_str=fields_str_to_save)
+    check_metadata_fields_expected_present(dict,fields_to_save)
+    dict[:Filename] = fn
+    if haskey(dict,:airmass) && typeof(dict[:airmass]) == String
+       dict[:airmass] = parse(Float64,dict[:airmass])
+    end
+    dict[:wfile] = joinpath(fn[1:end-33],"waves",dict[:wfile])
+    return dict
+end
+
+function add_metadata_from_fits!(df::DataFrame, fn::String)
+    println("# Reading metadata from ", fn)
+    metadata_keep = read_metadata(fn)
+    #reject file from df if it is a calibration (only keep spectra)
+    #if metadata_keep[:target] != "Solar" return df end
+    #calibration_target_substrings = ["Etalon","LFC","ThArS","LDLS","FlatBB"]
+    #if any(map(ss->contains( metadata_keep[:target],ss),calibration_target_substrings)) 
+    #   return df
+    #end
+
+    for k in propertynames(df)
+        if typeof(metadata_keep[k]) == Missing continue end
+        if typeof(metadata_keep[k]) == Nothing
+            metadata_keep[k] = missing
+            continue
+        end
+        if (eltype(df[!,k]) <: Union{Real,Union{<:Real,Missing}}) && !(typeof(metadata_keep[k]) <: Union{Real,Union{<:Real,Missing}})
+            metadata_keep[k] = missing
+        end
+    end
+    #println("metadata_keep = ", metadata_keep)
+    if haskey(metadata_keep,:airmass) && typeof(metadata_keep[:airmass]) == String
+       metadata_keep[:airmass] = parse(Float64,metadata_keep[:airmass])
+    end
+
+    metadata_keep[:wfile] = joinpath(fn[1:end-33],"waves",metadata_keep[:wfile])
+
+    push!(df, metadata_keep)
+    return df
+end
+
+
+
+
+""" Read HARPS wavelength file.
+    Note:HARPS wavelengths are stored in Float-32 format, which reduces wavelength precision.
+    This function restores that precision using a smooth polynomial fit """
+function read_λ(fn::String)
+    λ_fits = FITS(fn)
+    λ_original = convert(Array{Float64,2},FITSIO.read(λ_fits[1]))
+    nx, norder = size(λ_original)
+    λ_new = zeros(nx,norder)
+    for i in 1:norder
+        waves_fit = Polynomials.fit(1:nx,λ_original[:,i],2)
+        λ_new[:,i] = waves_fit.(1:nx)
+    end
+    return λ_new
+end
+
+function read_λ(fn::String, orders_to_read::AR ) where  { AR<:AbstractRange }
+    λ_fits = FITS(fn)
+    λ_original = convert(Array{Float64,2},FITSIO.read(λ_fits[1],:,orders_to_read))
+    nx, norder = size(λ_original)
+    λ_new = zeros(nx,norder)
+    for i in 1:norder
+        waves_fit = Polynomials.fit(1:nx,λ_original[:,i],2)
+        λ_new[:,i] = waves_fit.(1:nx)
+    end
+    return λ_new
+end
+
+""" Read HARPS data from FITS file, and return in a Spectra2DBasic object."""
+function read_data   end
+
+function read_data(f::FITS, metadata::Dict{Symbol,Any} )
+    @assert length(f) >= 1
+    #@assert all(map(i->typeof(f[i]) <: FITSIO.ImageHDU,2:length(f)))
+    λ = read_λ(metadata[:wfile])
+    flux = FITSIO.read(f[1])
+    var = NaNMath.sqrt.(flux)
```

Comment:
That don't include flux uncertainties?

Reply:
Even if not, what is var getting sqrt of flux, rather than flux?

Reply:
fixed to simply take a copy of flux